### PR TITLE
[WB-5641 and WB-6016] Reduce Memory Footprint of Images In Tables

### PR DESCRIPTION
**Pull request by tssweeney**
https://wandb.atlassian.net/browse/WB-5641
https://wandb.atlassian.net/browse/WB-6016

Description
-----------

What does the PR do?

This PR frees the wandb.Image ram when adding Images to tables. It also puts the `.image` property behind a property accessor function so that it can re-load the data if needed by the user. As you can see by the following printouts, without the optimization, 20k rows of 3 images of size 32x32x3 bytes has a cost of about 154.6 MiB. 200k rows of the same hits an overflow issue! With the optimization, we see 17.2 MiB and 180.7 MiB respectively - about an order of magnitude improvement. We are still adding 10x more rows to each table test, and there are inherently other items in memory which take up space, so this linear growth is expected. However, the image data itself offers a massive reduction.

This problem was surfaced via a Kaggle user: https://weightsandbiases.slack.com/archives/C01ECAXNRTL/p1622486557151400 and repro: https://www.kaggle.com/timssweeney/rfcx-hear-and-see-the-species-with-w-b/edit. With this change, we are able to get about 40% more rows before running out of memory. However, in order to really allow the user to have as many rows as they want, we will need to free table memory itself

Testing on: https://colab.research.google.com/drive/1lRwM1BNGTfCEHNwZ1PynWWDmHkkMhcm2?usp=sharing#scrollTo=ZZqZ9WxRIj2c. The step 2 breaks the notebook - using this code, RAM is kept very small.

## Without Opt:
```
Completed 1501 rows
 17%|█▋        | 1540/8997 [14:14<1:06:05,  1.88it/s]
```

## With Opt:
Note: this includes a matplotlib optimization
```
Completed 5001 rows
 58%|█████▊    | 5198/8997 [56:25<1:08:13,  1.08s/it]
```

## Before Optimization:

```
Starting count 2
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     80.1 MiB     80.1 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     80.1 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     82.0 MiB      0.0 MiB           2       for _ in range(count):
    10     82.0 MiB      1.9 MiB           6           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 2 in 0.24834108352661133 seconds
Starting count 20
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     82.0 MiB     82.0 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     82.0 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     82.5 MiB      0.0 MiB          11       for _ in range(count):
    10     82.5 MiB      0.5 MiB          60           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 20 in 0.07385706901550293 seconds
Starting count 200
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     82.5 MiB     82.5 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     82.5 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     83.4 MiB      0.0 MiB         101       for _ in range(count):
    10     83.4 MiB      0.9 MiB         600           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 200 in 0.6641268730163574 seconds
Starting count 2000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     83.4 MiB     83.4 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     83.4 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     85.8 MiB     -0.0 MiB        1001       for _ in range(count):
    10     85.8 MiB      2.2 MiB        6000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 2000 in 6.578912019729614 seconds
Starting count 20000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     85.3 MiB     85.3 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     85.3 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9    102.6 MiB      0.0 MiB       10001       for _ in range(count):
    10    102.6 MiB     17.2 MiB       60000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 20000 in 64.03094482421875 seconds
Starting count 200000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     86.3 MiB     86.3 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     86.3 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9    267.0 MiB      0.0 MiB      100001       for _ in range(count):
    10    267.0 MiB    180.7 MiB      600000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 200000 in 705.2537281513214 seconds
------------------------------------------------------------
[09:07:39]: ~/workspace/wandb-client/standalone_tests (tables/minimize_mem_footprint)
» python -m memory_profiler table_mem_test.py                         timothysweeney@Timothys-MacBook-Pro 
Starting count 2
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     79.9 MiB     79.9 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     79.9 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     81.9 MiB      0.0 MiB           2       for _ in range(count):
    10     81.9 MiB      2.0 MiB           6           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 2 in 0.25966310501098633 seconds
Starting count 20
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     81.9 MiB     81.9 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     81.9 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     82.5 MiB      0.0 MiB          11       for _ in range(count):
    10     82.5 MiB      0.6 MiB          60           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 20 in 0.06778979301452637 seconds
Starting count 200
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     82.5 MiB     82.5 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     82.5 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     84.4 MiB      0.0 MiB         101       for _ in range(count):
    10     84.4 MiB      1.9 MiB         600           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 200 in 0.7008109092712402 seconds
Starting count 2000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     84.4 MiB     84.4 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     84.4 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9    100.7 MiB      0.0 MiB        1001       for _ in range(count):
    10    100.7 MiB     16.3 MiB        6000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 2000 in 6.2767698764801025 seconds
Starting count 20000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6    100.7 MiB    100.7 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8    100.7 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9    255.3 MiB      0.0 MiB       10001       for _ in range(count):
    10    255.3 MiB    154.6 MiB       60000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 20000 in 63.144949197769165 seconds
Starting count 200000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6    173.6 MiB    173.6 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8    173.6 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9   1798.5 MiB -10522.4 MiB      100001       for _ in range(count):
    10   1798.5 MiB -61509.9 MiB      600000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])
```

## After Optimization:
```
Starting count 2
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     80.1 MiB     80.1 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     80.1 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     82.0 MiB      0.0 MiB           2       for _ in range(count):
    10     82.0 MiB      1.9 MiB           6           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 2 in 0.24834108352661133 seconds
Starting count 20
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     82.0 MiB     82.0 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     82.0 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     82.5 MiB      0.0 MiB          11       for _ in range(count):
    10     82.5 MiB      0.5 MiB          60           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 20 in 0.07385706901550293 seconds
Starting count 200
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     82.5 MiB     82.5 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     82.5 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     83.4 MiB      0.0 MiB         101       for _ in range(count):
    10     83.4 MiB      0.9 MiB         600           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 200 in 0.6641268730163574 seconds
Starting count 2000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     83.4 MiB     83.4 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     83.4 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9     85.8 MiB     -0.0 MiB        1001       for _ in range(count):
    10     85.8 MiB      2.2 MiB        6000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 2000 in 6.578912019729614 seconds
Starting count 20000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     85.3 MiB     85.3 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     85.3 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9    102.6 MiB      0.0 MiB       10001       for _ in range(count):
    10    102.6 MiB     17.2 MiB       60000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 20000 in 64.03094482421875 seconds
Starting count 200000
Filename: table_mem_test.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     6     86.3 MiB     86.3 MiB           1   @profile
     7                                         def main(count, size=(32, 32, 3)):
     8     86.3 MiB      0.0 MiB           1       table = wandb.Table(columns=["img_1", "img_2", "img_3"])
     9    267.0 MiB      0.0 MiB      100001       for _ in range(count):
    10    267.0 MiB    180.7 MiB      600000           table.add_data(*[wandb.Image(np.random.randint(255, size=size)) for _ in range(3)])


Completed count 200000 in 705.2537281513214 seconds
```

Testing
-------

How was this PR tested?

### CODE REVIEW

**Review by shawnlewis — APPROVED** [review]

File: wandb/data_types.py [hunk 1]
```diff
@@ -411,6 +411,10 @@ def add_data(self, *data):
         # Add the new data
         self.data.append(data)
 
+        for item in data:
+            if isinstance(item, Image):
+                item._free_ram()
```
Comment (vanpelt) [inline wandb/data_types.py:416]: It's a bit unfortunate we would need to do this for any data types that store data in memory.  Could we move this into the `Image`?  I was thinking we could reset the image in `to_data_array` assuming that's the last time we need attributes about the data itself.  This way we would only have one place to update other data types.

**Review by vanpelt — CHANGES_REQUESTED** [review]

File: wandb/data_types.py (hunk 1 above)
Comment (tssweeney) [reply wandb/data_types.py:416]: Hmm - this is a good recommendation. In the next commit I removed this Table-specific check and put the `_free_ram` at the end of the constructor. The `to_data_array` is a user-facing function and rarely called. This way the default state of an Image is not to hold an in-memory copy of the data, however, after any access to `Image.image`, a copy is retained. We can move this pattern to other Media types easily as you pointed out. LMK what you think.

**Review by vanpelt — APPROVED** [review]
Looks good.  The only cavaet is if someone calls `to_data_array` all the images will be loaded back into memory but if this is rare, that should be ok.

**Comment by tssweeney** [comment]
> Looks good. The only cavaet is if someone calls `to_data_array` all the images will be loaded back into memory but if this is rare, that should be ok.

Fair enough (: - I added to to the end of that function as well.

**Pull request merged**